examples/microwave-oven-app/microwave-oven-common/include/microwave-oven-device.h

```diff
@@ -0,0 +1,235 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/clusters/microwave-oven-control-server/microwave-oven-control-server.h>
+#include <app/clusters/mode-base-server/mode-base-server.h>
+#include <app/clusters/operational-state-server/operational-state-server.h>
+#include <app/util/af-enums.h>
+#include <app/util/config.h>
+#include <cstring>
+#include <protocols/interaction_model/StatusCode.h>
+#include <utility>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+/**
+ * set default value for the optional attributes
+ */
+constexpr uint8_t kDefaultMinPower  = 10u;
+constexpr uint8_t kDefaultMaxPower  = 100u;
+constexpr uint8_t kDefaultPowerStep = 10u;
+
+constexpr uint8_t ModeNormal  = 0;
+constexpr uint8_t ModeDefrost = 1;
+
+class ExampleMicrowaveOvenDevice : public MicrowaveOvenControl::Delegate,
+                                   public ModeBase::Delegate,
+                                   public OperationalState::Delegate
+{
+
+public:
+    /**
+     * This class is responsible for initialising all the microwave oven device clusters and managing the interactions between them
+     * as required by the specific "business logic". See the state machine diagram.
+     * @param aClustersEndpoint The endpoint ID where all the microwave oven clusters exist.
+     */
+    explicit ExampleMicrowaveOvenDevice(EndpointId aClustersEndpoint) :
+        mMicrowaveOvenControlInstance(this, aClustersEndpoint, MicrowaveOvenControl::Id, &mOperationalStateInstance,
+                                      &mMicrowaveOvenModeInstance),
+        mOperationalStateInstance(this, aClustersEndpoint, OperationalState::Id),
+        mMicrowaveOvenModeInstance(this, aClustersEndpoint, MicrowaveOvenMode::Id, 0)
```

Comment:
This initializer order passes pointers to not-yet-constructed objects to mMicrowaveOvenControlInstance.

And in the destructor the things it points to will be destroyed before it is destroyed, no?  Please fix the ordering here (both the order of the class members and the order of initializers).